### PR TITLE
Adding mux

### PR DESCRIPTION
**Pull request by Tzanccc**
Merging long range state machine and mux code into testing-pi.

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -82,9 +82,12 @@
         "variant": "cpp"
     },
     "python.autoComplete.extraPaths": [
+        "/home/arthur/WRover_Software/devel/lib/python3/dist-packages",
```
Comment (bennowotny) [inline .vscode/settings.json:85]: Relative path?
```suggestion
        "${workspaceFolder}/devel/lib/python3/dist-packages",
```

File: src/wr_logic_ai/srv/EmptySrv.srv
Comment (bennowotny) [inline src/wr_logic_ai/srv/EmptySrv.srv:1]: This already exists as a standard service type.  This custom type should not exist.

File: src/wr_logic_ai/CMakeLists.txt [hunk 2]
```diff
@@ -51,31 +52,31 @@ find_package(catkin REQUIRED COMPONENTS
...
 
 ## Generate services in the 'srv' folder
-# add_service_files(
-#   FILES
-#   Service1.srv
-#   Service2.srv
-# )
+add_service_files(
+   FILES
+   EmptySrv.srv
```
Comment (bennowotny) [inline src/wr_logic_ai/CMakeLists.txt:63]: See notes on file.

File: src/wr_logic_ai/nodes/long_range_action_server.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
+#!/usr/bin/env python3
+
+import rospy
+import actionlib
+from wr_logic_ai.msg import LongRangeAction, LongRangeGoal
+from obstacle_avoidance import update_target 
+
+# TODO: check timeout time length validity
+LONG_RANGE_TIMEOUT_TIME = rospy.Duration(5)
```
Comment (bennowotny) [inline src/wr_logic_ai/nodes/long_range_action_server.py:9]: Maybe slightly higher for testing.
```suggestion
LONG_RANGE_TIMEOUT_TIME = rospy.Duration(10)
```

File: src/wr_logic_ai/nodes/state_machine/state_machine.py [hunk 4]
```diff
@@ -1,95 +1,127 @@
+#!/usr/bin/env python
+
 from statemachine import StateMachine, State
-from coordinate_manager import CoordinateManager
-from wr_logic_ai.msg import TargetMsg
-from wr_logic_ai.msg import NavigationStateMsg
+from wr_logic_ai.coordinate_manager import CoordinateManager
+from wr_logic_ai.msg import NavigationState, LongRangeAction, LongRangeGoal
+from wr_logic_ai.srv import EmptySrv
```
Comment (bennowotny) [inline src/wr_logic_ai/nodes/state_machine/state_machine.py:6]: Use `std_srv`

File: src/wr_logic_ai/nodes/state_machine/state_machine.py [hunk 5]
```diff
@@ -1,95 +1,127 @@
+#!/usr/bin/env python
+
 from statemachine import StateMachine, State
```
Comment (bennowotny) [inline src/wr_logic_ai/nodes/state_machine/state_machine.py:3]: Use the `requirements.txt` to install this package (`python-statemachine==2.0.0`)

File: src/wr_logic_ai/nodes/state_machine/state_machine.py [hunk 6]
```diff
@@ -1,95 +1,127 @@
...
-            #self.pub_nav = rospy.Publisher('/target_coord', TargetMsg, queue_size=1)
-            #rospy.spin()
+            print(self._mgr.get_coordinate())  
+            self.timer = rospy.Timer(rospy.Duration(0.2), lambda _: self.mux_pub.publish(self.mux_long_range))           
+            client = actionlib.SimpleActionClient("LongRangeActionServer", LongRangeAction)
+            client.wait_for_server()
+            goal = LongRangeGoal(target_lat = self._mgr.get_coordinate()["lat"], target_long = self._mgr.get_coordinate()["long"])
+            client.send_goal(goal)
+            client.wait_for_result()
+            as_state = client.get_state()
```
Comment (bennowotny) [inline src/wr_logic_ai/nodes/state_machine/state_machine.py:79]: Use `send_goal_and_wait` (https://docs.ros.org/en/diamondback/api/actionlib/html/classactionlib_1_1simple__action__client_1_1SimpleActionClient.html#a3de1204015fe23bfd1b47b7fe164df34)

File: src/wr_logic_ai/nodes/state_machine/wait_for_user_input.py [hunk 7]
```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/env python
+
+import rospy
+from wr_logic_ai.srv import EmptySrv, EmptySrvResponse
```
Comment (bennowotny) [inline src/wr_logic_ai/nodes/state_machine/wait_for_user_input.py:4]: Use standard empty service.

File: src/wr_logic_ai/src/wr_logic_ai/coordinate_manager.py [hunk 8]
```diff
@@ -11,27 +11,24 @@ def __init__(self) -> None:
...
     @staticmethod
     def previous_line():
-        print(CoordinateManager.ptr)
-        CoordinateManager.ptr = CoordinateManager.ptr - 1
-        print(CoordinateManager.ptr)
+        CoordinateManager.ptr = max(0,CoordinateManager.ptr - 1)
 
     @staticmethod
-    def next_line():
+    def next_line() -> bool:
```
Comment (bennowotny) [inline src/wr_logic_ai/src/wr_logic_ai/coordinate_manager.py:21]: Better name: `next_coordinate`

File: src/wr_logic_ai/src/wr_logic_ai/coordinate_manager.py [hunk 9]
```diff
@@ -11,27 +11,24 @@ def __init__(self) -> None:
 
     @staticmethod
     def get_coordinate():
-        print(CoordinateManager.ptr)
         return CoordinateManager.coordinates[CoordinateManager.ptr]
         
     @staticmethod
     def previous_line():
```
Comment (bennowotny) [inline src/wr_logic_ai/src/wr_logic_ai/coordinate_manager.py:17]: Better name: `previous_coordinate`

**Review by bennowotny — CHANGES_REQUESTED** [review]

File: .vscode/settings.json (hunk 1 above)
Comment (Tzanccc) [reply .vscode/settings.json]: Addressed in [277b00a](https://github.com/WisconsinRobotics/WRover_Software/pull/48/commits/277b00a47c8537aba6ba2aeeb6c09d9b5ea4d3eb).

File: src/wr_logic_ai/srv/EmptySrv.srv
Comment (Tzanccc) [reply src/wr_logic_ai/srv/EmptySrv.srv:1]: Addressed in [277b00a](https://github.com/WisconsinRobotics/WRover_Software/pull/48/commits/277b00a47c8537aba6ba2aeeb6c09d9b5ea4d3eb).

File: src/wr_logic_ai/CMakeLists.txt (hunk 2 above)
Comment (Tzanccc) [reply src/wr_logic_ai/CMakeLists.txt]: Addressed in [277b00a](https://github.com/WisconsinRobotics/WRover_Software/pull/48/commits/277b00a47c8537aba6ba2aeeb6c09d9b5ea4d3eb).

File: src/wr_logic_ai/nodes/long_range_action_server.py (hunk 3 above)
Comment (Tzanccc) [reply src/wr_logic_ai/nodes/long_range_action_server.py]: Addressed in [277b00a](https://github.com/WisconsinRobotics/WRover_Software/pull/48/commits/277b00a47c8537aba6ba2aeeb6c09d9b5ea4d3eb).

File: src/wr_logic_ai/nodes/state_machine/state_machine.py (hunk 4 above)
Comment (Tzanccc) [reply src/wr_logic_ai/nodes/state_machine/state_machine.py]: Addressed in [277b00a](https://github.com/WisconsinRobotics/WRover_Software/pull/48/commits/277b00a47c8537aba6ba2aeeb6c09d9b5ea4d3eb).

File: src/wr_logic_ai/nodes/state_machine/state_machine.py (hunk 5 above)
Comment (Tzanccc) [reply src/wr_logic_ai/nodes/state_machine/state_machine.py:3]: Addressed in [277b00a](https://github.com/WisconsinRobotics/WRover_Software/pull/48/commits/277b00a47c8537aba6ba2aeeb6c09d9b5ea4d3eb).

File: src/wr_logic_ai/nodes/state_machine/state_machine.py (hunk 6 above)
Comment (Tzanccc) [reply src/wr_logic_ai/nodes/state_machine/state_machine.py]: Addressed in [277b00a](https://github.com/WisconsinRobotics/WRover_Software/pull/48/commits/277b00a47c8537aba6ba2aeeb6c09d9b5ea4d3eb).

File: src/wr_logic_ai/nodes/state_machine/wait_for_user_input.py (hunk 7 above)
Comment (Tzanccc) [reply src/wr_logic_ai/nodes/state_machine/wait_for_user_input.py]: Addressed in [277b00a](https://github.com/WisconsinRobotics/WRover_Software/pull/48/commits/277b00a47c8537aba6ba2aeeb6c09d9b5ea4d3eb).

File: src/wr_logic_ai/src/wr_logic_ai/coordinate_manager.py (hunk 8 above)
Comment (Tzanccc) [reply src/wr_logic_ai/src/wr_logic_ai/coordinate_manager.py]: Addressed in [277b00a](https://github.com/WisconsinRobotics/WRover_Software/pull/48/commits/277b00a47c8537aba6ba2aeeb6c09d9b5ea4d3eb).

File: src/wr_logic_ai/src/wr_logic_ai/coordinate_manager.py (hunk 9 above)
Comment (Tzanccc) [reply src/wr_logic_ai/src/wr_logic_ai/coordinate_manager.py]: Addressed in [277b00a](https://github.com/WisconsinRobotics/WRover_Software/pull/48/commits/277b00a47c8537aba6ba2aeeb6c09d9b5ea4d3eb).

File: src/wr_logic_ai/launch/long_range.launch [hunk 10]
```diff
@@ -1,11 +1,15 @@
 <launch>
   <include file="$(find wr_entry_point)/launch/comp/comp_init.launch" />
   <include file="$(find wr_hsi_sensing)/launch/hw_test.launch" />
```
Comment (bennowotny) [inline src/wr_logic_ai/launch/long_range.launch:3]: move to statemachine.launch

File: src/wr_logic_ai/launch/long_range.launch [hunk 11]
```diff
@@ -1,11 +1,15 @@
...
-  <!-- <node pkg="wr_logic_ai" name="testData" type="inputFakeData.py" output="screen"/> -->
+
+  <node pkg="wr_logic_ai" name="obstacle_avoidance" type="obstacle_avoidance.py" output="screen" long_range_mux_arg="$(arg long_range_output)"/>
+  
   <include file="$(find rplidar_ros)/launch/rplidar.launch" />
+  
   <node pkg="rviz" name="$(anon rviz)" type="rviz" output="screen" machine="base_station"
     args="-d $(find wr_logic_ai)/launch/obstacle_avoidance.rviz" />
+  
   <node name="control" pkg="wr_control_drive_ds" type="DriveTrainControlSystem" machine="wrover" />
```
Comment (bennowotny) [inline src/wr_logic_ai/launch/long_range.launch:12]: move to statemachine.launch

File: src/wr_logic_ai/launch/long_range.launch [hunk 12]
```diff
@@ -1,11 +1,15 @@
 <launch>
   <include file="$(find wr_entry_point)/launch/comp/comp_init.launch" />
   <include file="$(find wr_hsi_sensing)/launch/hw_test.launch" />
-  <node pkg="wr_logic_ai" name="obstacle_avoidance" type="obstacle_avoidance.py" output="screen" />
-  <!-- <node pkg="wr_logic_ai" name="testData" type="inputFakeData.py" output="screen"/> -->
+
+  <node pkg="wr_logic_ai" name="obstacle_avoidance" type="obstacle_avoidance.py" output="screen" long_range_mux_arg="$(arg long_range_output)"/>
+  
   <include file="$(find rplidar_ros)/launch/rplidar.launch" />
```
Comment (bennowotny) [inline src/wr_logic_ai/launch/long_range.launch:7]: move to statemachine.launch

File: src/wr_logic_ai/launch/mux_select.launch [hunk 13]
```diff
@@ -1,17 +1,19 @@
...
-    
     <arg name="mux_name" default="nav_mux"/>
     <arg name="long_range_output" default="/long_range_nav"/>
     <arg name="short_range_output" default="/short_range_nav"/>
-    <include file="$(find wr_logic_ai)/launch/long_range.launch")>
-        <arg name="long_range_output" default="/long_range_nav"/>
+
+    <include file="$(find wr_logic_ai)/launch/long_range.launch)">
+        <arg name="long_range_output" value = "$(arg long_range_output)" default="/long_range_nav"/>
     </include>
```
Comment (bennowotny) [inline src/wr_logic_ai/launch/mux_select.launch]: This should be started by statemachine.launch

File: src/wr_logic_ai/launch/state_machine.launch [hunk 14]
```diff
@@ -2,4 +2,5 @@
   <node pkg="wr_logic_ai" name="wait_for_user_input" type="wait_for_user_input.py" output="screen"/>
   <node pkg="wr_logic_ai" name="long_range_action_server" type="long_range_action_server.py" output="screen"/>
   <node pkg="wr_logic_ai" name="nav_state_machine_driver" type="state_machine.py" output="screen"/>
+  <include file="$(find wr_logic_ai)/launch/mux_select.launch)">
```
Comment (bennowotny) [inline src/wr_logic_ai/launch/state_machine.launch:5]: Pass args to set the actual topic names for the mux

File: src/wr_logic_ai/launch/long_range.launch [hunk 15]
```diff
@@ -1,11 +1,15 @@
 <launch>
   <include file="$(find wr_entry_point)/launch/comp/comp_init.launch" />
   <include file="$(find wr_hsi_sensing)/launch/hw_test.launch" />
-  <node pkg="wr_logic_ai" name="obstacle_avoidance" type="obstacle_avoidance.py" output="screen" />
-  <!-- <node pkg="wr_logic_ai" name="testData" type="inputFakeData.py" output="screen"/> -->
+
+  <node pkg="wr_logic_ai" name="obstacle_avoidance" type="obstacle_avoidance.py" output="screen" long_range_mux_arg="$(arg long_range_output)"/>
```
Comment (bennowotny) [inline src/wr_logic_ai/launch/long_range.launch:5]:
```suggestion
  <node pkg="wr_logic_ai" name="obstacle_avoidance" type="obstacle_avoidance.py" output="screen">
    <param name="motor_speeds" value="$(arg long_range_output)" />
  </node>
```

File: src/wr_logic_ai/nodes/obstacle_avoidance.py [hunk 16]
```diff
@@ -39,11 +39,11 @@ def initialize() -> None:
...
 
-# Initialize node
-rospy.init_node('nav_autonomous', anonymous=False)
-
+    # Initialize node
+    rospy.init_node('nav_autonomous', anonymous=False)
+    
     # Publisher
-    drive_pub = rospy.Publisher('/control/drive_system/cmd', DriveTrainCmd, queue_size=1)
+    drive_pub = rospy.Publisher(rospy.get_param('~long_range_mux_arg'), DriveTrainCmd, queue_size=1)
```
Comment (bennowotny) [inline src/wr_logic_ai/nodes/obstacle_avoidance.py:46]:
```suggestion
    drive_pub = rospy.Publisher(rospy.get_param('~motor_speeds'), DriveTrainCmd, queue_size=1)
```

**Review by bennowotny — CHANGES_REQUESTED** [review]

File: src/wr_logic_ai/launch/long_range.launch [hunk 17]
```diff
@@ -1,15 +1,10 @@
...
-  <node pkg="wr_logic_ai" name="obstacle_avoidance" type="obstacle_avoidance.py" output="screen" long_range_mux_arg="$(arg long_range_output)"/>
-  
-  <include file="$(find rplidar_ros)/launch/rplidar.launch" />
-  
   <node pkg="rviz" name="$(anon rviz)" type="rviz" output="screen" machine="base_station"
     args="-d $(find wr_logic_ai)/launch/obstacle_avoidance.rviz" />
   
-  <node name="control" pkg="wr_control_drive_ds" type="DriveTrainControlSystem" machine="wrover" />
-  
   <include file="$(find wrevolution)/launch/ros_control_test.launch" />
```
Comment (bennowotny) [inline src/wr_logic_ai/launch/long_range.launch:9]: This should go to statemachine.launch.

File: src/wr_logic_ai/launch/mux_select.launch [hunk 18]
```diff
@@ -1,12 +1,4 @@
 <launch>
-    <arg name="mux_name" default="nav_mux"/>
-    <arg name="long_range_output" default="/long_range_nav"/>
-    <arg name="short_range_output" default="/short_range_nav"/>
```
Comment (bennowotny) [inline src/wr_logic_ai/launch/mux_select.launch]: The args should stay so you can pass them in.

**Review by bennowotny — CHANGES_REQUESTED** [review]

**Review by bennowotny — APPROVED** [review]

**Pull request merged**